add command palette (re-pr)

plyght opened the pull request:
This pull request includes several changes to the `settings` and `sidebar` components, as well as the addition of a new `CommandPaletteProvider` to the application. The most important changes focus on improving the user interface and refactoring the code for better readability and maintainability.

### User Interface Improvements:
* [`app/(routes)/settings/general/page.tsx`](diffhunk://#diff-e489cf82b0075442d1dc78c7aea2d02fc9ec450ca2c2c2b2d29920fcd6891c53R173-R195): Added a new form item for collapsing the sidebar and moved the "Sign Out" button to a new `SettingsCard` under the "Account" section.
* [`app/layout.tsx`](diffhunk://#diff-eca96d2c09f31517696a26e1d0be4070e1fbab02831481bed006e275741d030bL42-R43): Wrapped the `children` component with the new `CommandPaletteProvider` to enable command palette functionality throughout the application.

### Code Refactoring:
* [`app/(routes)/settings/general/page.tsx`](diffhunk://#diff-e489cf82b0075442d1dc78c7aea2d02fc9ec450ca2c2c2b2d29920fcd6891c53L38-R39): Removed the unused `useRouter` import and the associated code for handling sign-out redirection. [[1]](diffhunk://#diff-e489cf82b0075442d1dc78c7aea2d02fc9ec450ca2c2c2b2d29920fcd6891c53L38-R39) [[2]](diffhunk://#diff-e489cf82b0075442d1dc78c7aea2d02fc9ec450ca2c2c2b2d29920fcd6891c53L66-R62)
* [`components/context/sidebar-context.tsx`](diffhunk://#diff-dffde709591899be86a3721d5c7e2ad4a0d82022ca27b4d473976514cf07ba28L59-L60): Removed redundant comments and simplified the code for handling sidebar state and keyboard shortcuts. [[1]](diffhunk://#diff-dffde709591899be86a3721d5c7e2ad4a0d82022ca27b4d473976514cf07ba28L59-L60) [[2]](diffhunk://#diff-dffde709591899be86a3721d5c7e2ad4a0d82022ca27b4d473976514cf07ba28L71-L90) [[3]](diffhunk://#diff-dffde709591899be86a3721d5c7e2ad4a0d82022ca27b4d473976514cf07ba28L103-L104)
* [`components/ui/app-sidebar.tsx`](diffhunk://#diff-3bd99b5941e8cda4d4b2000581a670faf21ad68da058c83e98165bc749637c71L51-R56): Added the `isMobile` state to conditionally hide the sidebar on mobile devices.

### New Features:
* [`components/ui/command-palette.tsx`](diffhunk://#diff-5dac0c6487c62108ea83cc024d1368a7e328dee23dee1c3e118f37602e977c14R1-R145): Introduced a new `CommandPaletteProvider` component that provides a command palette for quick navigation and actions within the application. This includes keyboard shortcuts for opening the command palette and composing a new mail.

### Other Changes:
* [`app/(routes)/settings/layout.tsx`](diffhunk://#diff-6ae6f0e492918a2d4b497d61f62e65b52ea568e1b3a6dc4c815e8047e9e3eb88L20-L22): Removed commented-out code related to media queries.
* [`components/ui/app-sidebar.tsx`](diffhunk://#diff-3bd99b5941e8cda4d4b2000581a670faf21ad68da058c83e98165bc749637c71L10-R10): Added the `useIsMobile` import to check for mobile devices.

plyght commented:
> Conflicts please @plyght

@MrgSub Should be fixed!

plyght commented:
@MrgSub should finally work.

plyght commented:
# Updated PR Summary: 

This pull request introduces a new Command Palette feature and integrates it into the application's layout. The most important changes include adding a new `CommandPaletteProvider` component, updating the layout to include this provider, and implementing the Command Palette functionality.

### Command Palette Integration:

* [`app/layout.tsx`](diffhunk://#diff-eca96d2c09f31517696a26e1d0be4070e1fbab02831481bed006e275741d030bR1): Imported `CommandPaletteProvider` and wrapped the children components with it inside the `Suspense` component, providing a fallback loading message. [[1]](diffhunk://#diff-eca96d2c09f31517696a26e1d0be4070e1fbab02831481bed006e275741d030bR1) [[2]](diffhunk://#diff-eca96d2c09f31517696a26e1d0be4070e1fbab02831481bed006e275741d030bL39-L42)

### Command Palette Implementation:

* [`components/ui/command-palette.tsx`](diffhunk://#diff-5dac0c6487c62108ea83cc024d1368a7e328dee23dee1c3e118f37602e977c14R1-R145): Added the new `CommandPaletteProvider` component, which manages the state and behavior of the Command Palette. This includes handling keyboard shortcuts, defining commands, and rendering the Command Palette UI.

Praashh reviewed (changes requested):
it looks good to me, at last can you please add a loom.

plyght commented:
@Praashh 

https://github.com/user-attachments/assets/14555ac3-6639-42b6-aff2-fb8b05f3b7f9

Praashh reviewed (approved):
LGTM!!

ahmetskilinc commented:
how did the .github files end up in here?

plyght commented:
@ahmetskilinc I manually copied them from the main repo (Mail-0/Mail-0) to my fork (plyght/Mail0-plyght)

plyght commented:
## Updated PR Summary w/ Updated Video:
This pull request includes several changes to improve the command palette functionality, refactor the shortcuts configuration, and enhance the overall user experience. The most important changes include the introduction of the `CommandPaletteProvider`, refactoring the shortcuts configuration, and updating the `CommandMenu` component to use the new command palette context.

### Command Palette Enhancements:
* [`components/ui/command-palette.tsx`](diffhunk://#diff-5dac0c6487c62108ea83cc024d1368a7e328dee23dee1c3e118f37602e977c14R1-R197): Introduced the `CommandPaletteProvider` to manage the state and context of the command palette, including handling keyboard shortcuts and running commands.
* [`app/layout.tsx`](diffhunk://#diff-eca96d2c09f31517696a26e1d0be4070e1fbab02831481bed006e275741d030bR1): Wrapped the application with `CommandPaletteProvider` and added a fallback for the `Suspense` component to improve user experience during loading. [[1]](diffhunk://#diff-eca96d2c09f31517696a26e1d0be4070e1fbab02831481bed006e275741d030bR1) [[2]](diffhunk://#diff-eca96d2c09f31517696a26e1d0be4070e1fbab02831481bed006e275741d030bL39-L42)

### Shortcuts Configuration:
* [`config/shortcuts.ts`](diffhunk://#diff-da0a61b6e5e07496ee960c91cc54e124bd9687b3af069186d82adbff9a3ef8abR1-R24): Created a new configuration file for keyboard shortcuts to centralize the shortcuts definitions.
* [`app/(routes)/settings/shortcuts/page.tsx`](diffhunk://#diff-b887bd7aeeaa94ce806342ed26408cec07c5ea512641ed17b8622caedee6e3a0R1-R9): Refactored the shortcuts array to import from the new configuration file.

### Command Menu Updates:
* [`components/ui/command-menu.tsx`](diffhunk://#diff-8ea66a67f555579a739cb1ca1c29b90b0050455926b5fb0f9145b393f93a3856R3-R18): Updated the `CommandMenu` component to use the new `useOpenComposeModal` hook and refactored the command handling logic to improve code readability and maintainability. [[1]](diffhunk://#diff-8ea66a67f555579a739cb1ca1c29b90b0050455926b5fb0f9145b393f93a3856R3-R18) [[2]](diffhunk://#diff-8ea66a67f555579a739cb1ca1c29b90b0050455926b5fb0f9145b393f93a3856L13-L107)

### Minor Updates:
* [`components/mail/mail.tsx`](diffhunk://#diff-0493710e0b0bcdca6e25e291c9aace467df8f1e7f9b7cb17582db2230349b317L65-R65): Added the `router` dependency to the `useEffect` dependency array to ensure proper navigation functionality.


https://github.com/user-attachments/assets/51416808-60f0-45d3-820c-c862ed1b650e

ahmetskilinc commented on components/mail/mail.tsx:
```diff
@@ -62,7 +62,7 @@ export function Mail({ folder }: MailProps) {
       router.push("/login");
     }
-  }, [session?.user, isPending]);
+  }, [session?.user, isPending, router]);
```
why the change?

plyght replied:
apologies, wrong use case. wil remove.

ahmetskilinc commented on components/ui/command-menu.tsx:
can delete

plyght commented on components/ui/command-menu.tsx:
```diff
@@ -10,101 +26,119 @@ import {
               <CommandShortcut>⌘C</CommandShortcut>
             </CommandItem>
-            <CommandItem>
+            <CommandItem onSelect={() => runCommand(() => console.log("Create label"))}>
```
file deleted.

plyght commented on app/layout.tsx:
```diff
@@ -36,10 +37,10 @@ export default function RootLayout({
       <body className={cn(geistSans.variable, geistMono.variable, "antialiased")}>
         <Providers attribute="class" defaultTheme="dark" enableSystem disableTransitionOnChange>
-          <Suspense>
+          <Suspense fallback={<div>Loading...</div>}>
```
fixed.

plyght commented:
@ahmetskilinc see if the changes are to your liking.
